segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,204 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.impl.bookkeeper;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import io.pravega.common.AbstractTimer;
+import io.pravega.common.MathHelpers;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Manages the Write Queue Concurrency.
+ */
+@ThreadSafe
+class ConcurrencyManager {
+    //region Members
+
+    @VisibleForTesting
+    static final double SIGNIFICANT_DIFFERENCE = 0.25;
+    @VisibleForTesting
+    static final long UPDATE_FREQUENCY_MILLIS = 1000;
```

Comment:
I am concerned that this algorithm which maintains only a comparison between the last interval and the current one, may react too quickly. After all this is being run per-container, so there are likely tens of similar calculations being made on the same host and hundreds more from other hosts all ultimately contending on the same resources. It may work well for a single process single segment, but that does not guarantee that there won't be wild fluctuations in the real world, which cause performance to be erratic or that settle with very different values for different instances when they operating on the same resource resulting in containers being permanently fast and others permanently slow.

Reply:
I don't disagree with you. However at this point I have no way of verifying that or even developing an "empirical theory" based on observed behavior since I do not have access to a large cluster to experiment upon.

However, a "large" cluster should have a sufficiently large number of bookies to reduce chances of one ledger interfering with the performance of another. Based on observed behavior, even with one Bookie (hence one index log), writing to several ledgers in parallel provided a noticeable performance boost compared to writing to the same ledger in parallel (same degree of parallelism), which indicates that there are per-ledger locks that affect performance inside BK, but it is unlikely for a ledger to affect others given that we don't overly constraint the BK cluster.